synapseutils/sync.py

```diff
@@ -589,188 +594,138 @@ def _order_items(items):
-
-                # we acquire the semaphore to ensure that we aren't uploading more
-                # than our configured maximum number of files here at once. once we
-                # reach the limit we'll block here until one of the existing file uploads completes
-                self._file_semaphore.acquire()
-                trace.get_current_span().set_attributes(
-                    {"thread.id": threading.get_ident()}
                 )
-                future = self._executor.submit(
-                    self._upload_item,
-                    item,
-                    used,
-                    executed,
-                    finished_items,
-                    pending_provenance,
-                    dependency_condition,
-                    abort_event,
-                    progress,
-                    context.get_current(),
-                )
-                futures.append(future)
-
-            with dependency_condition:
-                if pending_provenance.has_pending():
-                    # skipped_items contains all the items that we couldn't upload
-                    # the previous time through the loop because they depended on
-                    # another item for provenance. wait until there at least one
-                    # those items finishes before continuing another time through the loop.
-                    if not abort_event.is_set():
-                        dependency_condition.wait_for(
-                            lambda: (
-                                pending_provenance.has_finished_provenance()
-                                or abort_event.is_set()
-                            )
-                        )
-
-                pending_provenance.reset_count()
-
-            ordered_items = skipped_items
+                initial_items_to_pop.append(key)
+        for key in initial_items_to_pop:
+            ordered_items.pop(key)
 
-        # all items have been submitted for upload
-
-        concurrent.futures.wait(futures, return_when=concurrent.futures.FIRST_EXCEPTION)
-        if abort_event.is_set():
-            # at least one item failed to upload
-            self._abort(futures)
+        while async_tasks:
+            done_tasks, pending_tasks = await asyncio.wait(
+                async_tasks, return_when=asyncio.FIRST_COMPLETED
+            )
+            async_tasks = pending_tasks
+            for completed_task in done_tasks:
+                try:
+                    stored_file = completed_task.result()
+                    finished_items[stored_file.path] = stored_file
+                    ordered_items.pop(stored_file.path, None)
+
+                    for key, activity_dependencies in ordered_items.items():
+                        if activity_dependencies and all(
+                            activity_dependency in finished_items
+                            if os.path.isfile(activity_dependency)
+                            else True
```

Comment:
Am I reading this wrong or would this evaluate to `True` no matter what?

Reply:
Not quite. In an activity you have either a `URL`, a `syn123` ID, or a reference to a file on disk.

In this case the check is attempting to verify:

For every dependencies that you have:

If it's a file -> Make sure that it has already been saved
If it is not a file -> Pass over the dependency as we don't have anything it could be waiting on

This is the important piece how I am determining when I can store a node from the Directed Acyclic Graph:
https://www.geeksforgeeks.org/topological-sorting/

I want to modify the code in `_order_items` to instead give back an order that is a bit easier than this logic here. The difficult part the code here solved was to make sure that all outbound edge nodes were saved before the starting node is saved.
https://github.com/Sage-Bionetworks/synapsePythonClient/blob/SYNPY-1356-refactor-sync-to-synapse/synapseutils/sync.py#L571